src/aws/flb_s3_local_buffer.c

```diff
@@ -414,3 +414,106 @@ int flb_remove_chunk_files(struct flb_local_chunk *c)
     }
     return ret;
 }
+
+/* Generic replace function for strings. */
+flb_sds_t replace_uri_tokens(const char* original_string, const char* current_word, const char* new_word) 
+{ 
+    flb_sds_t result; 
+    int i = 0;
+    int count = 0; 
+    int new_word_len = strlen(new_word); 
+    int old_word_len = strlen(current_word); 
+  
+    for (i = 0; original_string[i] != '\0'; i++) { 
+        if (strstr(&original_string[i], current_word) == &original_string[i]) { 
+            count++; 
+            i += old_word_len - 1; 
+        } 
+    }
+
+    result = flb_sds_create_size(i + count * (new_word_len - old_word_len) + 1);
+    if (!result) {
+        flb_errno();
+        flb_sds_destroy(result);
+        return NULL;
+    }
+
+    i = 0; 
+    while (*original_string) { 
+        if (strstr(original_string, current_word) == original_string) { 
+            strcpy(&result[i], new_word); 
+            i += new_word_len; 
+            original_string += old_word_len; 
+        } 
+        else
+            result[i++] = *original_string++; 
+    } 
+
+    result[i] = '\0'; 
+
+    return result; 
+} 
+  
+//* Constructs S3 object key as per the format. */
+flb_sds_t get_s3_key(const char *format, time_t time, char *tag)
+{
+    int i = 0;
+    char tag_idx[9];
+    char datestamp[40];
+    flb_sds_t tag_token;
+    flb_sds_t s3_key = NULL;
+    flb_sds_t time_format = NULL;
+    struct tm *gmt;
+    flb_sds_t tag_delimiter = ".";
```

Comment:
This should be a char *, not an SDS. You can't declare an SDS like this (even though SDS is an alias for char *) because there's some memory magic used with SDS strings: https://github.com/antirez/sds#how-sds-strings-work

Reply:
Or just a char (no pointer) since its a single character

Reply:
Also, make it a function argument- I want to implement the tag_delimiters option that I talked about that makes the tag delimiters user configurable.

Reply:
initialize *gmt = NULL as well?

Reply:
Done. Added tag delimiter as function argument and added the corresponding test for multi tag delimiters.